From 342ce6923188c19f7689fd024929e0380c5db2de Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 31 Mar 2022 21:58:08 +0300
Subject: [PATCH] [SW-77251] fit: gaudi3: add copy image function

This patch add device copy image function for gaudi3. This function
will be called from fit operations to copy images at desired location.

Change-Id: I752e45aa4d175e6b7b12bd7ee20d8e525c909a90
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 7f95521..66fe3c1 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -714,3 +714,17 @@
 	image_entry = (void *)(*(uint32_t *)arc_addr);
 	image_entry();
 }
+
+#define ARC_ADDR_MASK			0xFFFFFFFull
+void arcboot_copy_image(uint64_t dst, const void *buf, int len, __unused const char *arch)
+{
+	uint32_t arc_dst;
+
+	/*
+	 * ARC_ADDR_MASK trancates the address to the 256MB region accesible by
+	 * Arc. MSB of 0x2 points to the FW_MEM_REGION.
+	 */
+	arc_dst = ((uint32_t)dst & ARC_ADDR_MASK) | BIT(29);
+
+	memcpy((void *)arc_dst, buf, len);
+}
